File: sdks/dotnet/Usergrid.Sdk.Tests/ClientTests/ConnectionTests.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


using System.Collections.Generic;
using NSubstitute;
using NUnit.Framework;
using Usergrid.Sdk.Manager;
using Usergrid.Sdk.Model;

namespace Usergrid.Sdk.Tests.ClientTests
{
    [TestFixture]
    public class ConnectionTests
    {
        #region Setup/Teardown

        [SetUp]
        public void Setup()
        {
            _connectionManager = Substitute.For<IConnectionManager>();
            _client = new Client(null, null) {ConnectionManager = _connectionManager};
        }

        #endregion

        private IConnectionManager _connectionManager;
        private IClient _client;

        [Test]
        public void CreateConnectionShouldDelegateToConnectionManagerWithCorrectParameters()
        {
            var connection = new Connection();
            _client.CreateConnection(connection);

            _connectionManager.Received(1).CreateConnection(connection);
        }

        [Test]
        public void DeleteConnectionsShouldDelegateToConnectionManagerWithCorrectParameters()
        {
            var connection = new Connection();

            _client.DeleteConnection(connection);

            _connectionManager.Received(1).DeleteConnection(connection);
        }

        [Test]
        public void GetConnectionsOfSpecificTypeShouldDelegateToConnectionManagerWithCorrectParameters()
        {
            var connection = new Connection();
            var enityList = new List<UsergridEntity>();

            _connectionManager.GetConnections<UsergridEntity>(connection).Returns(enityList);

            IList<UsergridEntity> returnedEntityList = _client.GetConnections<UsergridEntity>(connection);

            _connectionManager.Received(1).GetConnections<UsergridEntity>(connection);
            Assert.AreEqual(enityList, returnedEntityList);
        }

        [Test]
        public void GetConnectionsShouldDelegateToConnectionManagerWithCorrectParameters()
        {
            var connection = new Connection();
            var enityList = new List<UsergridEntity>();
            _connectionManager.GetConnections(connection).Returns(enityList);

            IList<UsergridEntity> returnedEntityList = _client.GetConnections(connection);

            _connectionManager.Received(1).GetConnections(connection);
            Assert.AreEqual(enityList, returnedEntityList);
        }
    }
}

File: sdks/swift/Tests/CONNECTION_Tests.swift
//
//  CONNECTION_Tests.swift
//  UsergridSDK
//
//  Created by Robert Walsh on 10/5/15.
//
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  The ASF licenses this file to You
 * under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.  For additional information regarding
 * copyright in this work, please see the NOTICE file in the top level
 * directory of this distribution.
 *
 */

import XCTest
@testable import UsergridSDK

class CONNECTION_Tests: XCTestCase {

    let clientAuth = UsergridAppAuth(clientId: "b3U6THNcevskEeOQZLcUROUUVA", clientSecret: "b3U6RZHYznP28xieBzQPackFPmmnevU")
    private static let collectionName = "publicevent"

    override func setUp() {
        super.setUp()
        Usergrid.initSharedInstance(orgId:ClientCreationTests.orgId, appId: "sdk.demo")
    }

    override func tearDown() {
        Usergrid._sharedClient = nil
        super.tearDown()
    }

    func test_CLIENT_AUTH() {

        let authExpect = self.expectationWithDescription("\(#function)")
        Usergrid.authMode = .App
        Usergrid.authenticateApp(clientAuth) { auth,error in
            XCTAssertTrue(NSThread.isMainThread())
            XCTAssertNil(error)
            XCTAssertNotNil(Usergrid.appAuth)

            if let appAuth = Usergrid.appAuth {

                XCTAssertNotNil(appAuth.accessToken)
                XCTAssertNotNil(appAuth.expiry)

                Usergrid.GET(CONNECTION_Tests.collectionName) { (response) in
                    XCTAssertTrue(NSThread.isMainThread())
                    XCTAssertNotNil(response)
                    XCTAssertTrue(response.ok)
                    XCTAssertTrue(response.hasNextPage)
                    XCTAssertEqual(response.entities!.count, 10)

                    let entity = response.first!
                    let entityToConnect = response.entities![1]
                    XCTAssertEqual(entity.type, CONNECTION_Tests.collectionName)

                    entity.connect("likes", toEntity: entityToConnect) { (response) -> Void in
                        XCTAssertTrue(NSThread.isMainThread())
                        XCTAssertNotNil(response)
                        XCTAssertTrue(response.ok)

                        entity.getConnections(.Out, relationship: "likes", query:nil) { (response) -> Void in
                            XCTAssertTrue(NSThread.isMainThread())
                            XCTAssertNotNil(response)
                            XCTAssertTrue(response.ok)

                            let connectedEntity = response.first!
                            XCTAssertNotNil(connectedEntity)
                            XCTAssertEqual(connectedEntity.uuidOrName, entityToConnect.uuidOrName)

                            entity.disconnect("likes", fromEntity: connectedEntity) { (response) -> Void in
                                XCTAssertTrue(NSThread.isMainThread())
                                XCTAssertNotNil(response)
                                XCTAssertTrue(response.ok)

                                entity.getConnections(.Out, relationship: "likes", query:nil) { (response) -> Void in
                                    XCTAssertTrue(NSThread.isMainThread())
                                    XCTAssertNotNil(response)
                                    XCTAssertTrue(response.ok)
                                    authExpect.fulfill()
                                }
                            }
                        }
                    }
                }
            } else {
                authExpect.fulfill()
            }
        }
        self.waitForExpectationsWithTimeout(20, handler: nil)
    }
}
